internal/raft/append_entries.go

```diff
@@ -6,6 +6,52 @@ import "github.com/tomarrell/lbadd/internal/raft/message"
 // to the follower node. This function generates the response to be sent to the leader node.
 // This is the response to the contact by the leader to assert it's leadership.
 func AppendEntriesResponse(node *Node, req *message.AppendEntriesRequest) *message.AppendEntriesResponse {
+	success := true
+	leaderTerm := req.GetTerm()
+	nodePersistentState := node.PersistentState
+	nodeTerm := nodePersistentState.CurrentTerm
+	// return false if term < currentTerm
+	if nodeTerm > leaderTerm {
+		success = false
+	} else if req.GetPrevLogIndex() > node.VolatileState.CommitIndex {
+		// return false if msg Log Index is greater than node commit Index
+		success = false
+	} else if nodePersistentState.Log[req.PrevLogIndex].Term != req.GetPrevLogTerm() {
+		// return false if term of msg at PrevLogIndex doesn't match prev Log Term stored by Leader
+		success = false
+	}
+
+	if !success {
+		return &message.AppendEntriesResponse{
+			Term:    nodeTerm,
+			Success: success,
+		}
+	}
+
+	entries := req.GetEntries()
+	// if heartbeat, skip adding entries
+	if len(entries) > 0 {
+		nodePersistentState.mu.Lock()
+		if req.GetPrevLogIndex() < node.VolatileState.CommitIndex {
+			node.PersistentState.Log = node.PersistentState.Log[:req.GetPrevLogIndex()]
```

Comment:
Potential memory leak

https://stackoverflow.com/questions/55045402/memory-leak-in-golang-slice

Reply:
Few things that is written about it - 

1. append can cause a reallocation and copy of the underlying array, which allows the old underlying array to be garbage collected.
2. If you were alarmed by that "most of the time" bit, please don't:
the backing array might not be reused only when append() operating
on the slice pointing into that array is requested to *grow* the slice
beyond the size of the underlying array.

But if this is still a concern - I can assign empty struct for the truncated entries. That will be fine right ?

Reply:
One more thing, LogData will be persisted on system like a file so at that time this won't be concern. LogData has to be added or removed from a system file before applying the given state to machine that's how persistence storage will work

Reply:
there's a difference between an empty struct and a nil pointer, so no

Reply:
both can be collected by GC